File: KimChaeJung/프로그래머스/lv2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations

def isPrimeNumber(number):
    if number == 1 or number == 0:
        return False
    for d in range(2, round(number**(1/2)+1)):
        if number%d == 0:
            return False
    return True
        
def solution(numbers):
    answer = set([])
    numberList = list(map(int, list(numbers)))
    
    for n in range(1, len(numberList)+1):
        for permu in permutations(numberList, n):
            permutatedNumber = int("".join(str(per) for per in permu))
            if isPrimeNumber(permutatedNumber):
                answer.add(permutatedNumber)

    return len(answer)


# 코드 개선
from itertools import permutations

def isPrimeNumber(number):
    if number < 2:
        return False
    if number % 2 == 0 and number > 2:  
        return False
    for d in range(3, int(number**0.5) + 1, 2):  
        if number % d == 0:
            return False
    return True
        
def solution(numbers):
    answer = set()
    digitList = list(numbers)
    
    for n in range(1, len(digitList) + 1):
        for permu in permutations(digitList, n):
            primeCandidate = int("".join(permu))
            if isPrimeNumber(primeCandidate):
                answer.add(primeCandidate)

    return len(answer)

File: KimChaeJung/프로그래머스/lv2/42839. 소수 찾기/소수 찾기.js
function isPrimeNumber(number){
    if (number < 2) return false
    for (let i = 2; i <= Math.sqrt(number); i++){
        if (number%i == 0) return false
    }
    return true
}

function* permutations(arr, n) {
    if (n == 1) {
        for (let ele of arr) yield [ele]
    } else {
        for (let i = 0; i < arr.length; i++){
            let rest = [...arr.slice(0, i), ...arr.slice(i+1)]
            for (let permu of permutations(rest, n-1)){
                yield [arr[i], ...permu]
            }
        }
    }
}

function solution(numbers) {
    const answerSet = new Set([])
    const digitList = numbers.split('')
    
    for (let digitCount = 1; digitCount <= digitList.length; digitCount++){
        for (const permu of permutations(digitList, digitCount)){
            const targetNumber = +permu.join("")
            
            if (isPrimeNumber(targetNumber)){
                answerSet.add(targetNumber)
            }
        }
    }
    
    
    
    return answerSet.size;
}